nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -718,16 +718,140 @@ impl StringPrototype {
+
+            // 8. Let advanceBy be max(1, searchLength).
+            let advance_by = max(1, search_length);
+
+            // 9. Let matchPositions be a new empty List.
+            let mut match_positions: Vec<usize> = vec![];
+
+            // 10. Let position be StringIndexOf(s, searchString, 0).
+            let search_str = search_string.as_str(agent).to_owned();
+            let subject = s.as_str(agent).to_owned();
+
+            let mut position = if let Some(position) = subject.find(&search_str) {
+                position
+            } else {
+                // If position is not-found, return s.
+                return Ok(s.into_value());
+            };
+
+            match_positions.push(position);
+            position += advance_by;
+
+            // 11. Repeat, while position is not not-found,
+            loop {
+                if let Some(pos) = subject[position..].find(&search_str) {
+                    match_positions.push(position + pos);
+                    position += advance_by;
+                } else {
+                    break;
+                };
+            }
+
+            // 12. Let endOfLastMatch be 0.
+            let mut end_of_last_match = 0;
+
+            // 13. Let result be the empty String.
+            let mut result = "".to_string();
+
+            // 14. For each element p of matchPositions, do
+            for p in match_positions {
+                // a. Let preserved be the substring of string from endOfLastMatch to p.
+                let preserved = subject[end_of_last_match..p].to_owned();
+
+                // b. let replacement be ? ToString(? Call(replaceValue, undefined, « searchString, 𝔽(p), string »)).
+                let replacement = call_function(
+                    agent,
+                    functional_replace,
+                    Value::Undefined,
+                    Some(ArgumentsList(&[
+                        search_string.into_value(),
+                        Number::from(position as u32).into_value(),
+                        s.into_value(),
+                    ])),
+                )?;
+
+                // d. Set result to the string-concatenation of result, preserved, and replacement.
+                let replacement_str = replacement.to_string(agent)?;
+                let replacement_str = replacement_str.as_str(agent);
+
+                let concat = format!("{}{}", &preserved, &replacement_str);
+                result.push_str(&concat);
+                end_of_last_match = p + search_length;
+            }
+
+            // 15. If endOfLastMatch < the length of string, set result to the string-concatenation of result and the substring of string from endOfLastMatch.
+            if end_of_last_match < subject.len() {
+                let preserved = subject[end_of_last_match..].to_owned();
+                result.push_str(&preserved);
+            }
+
+            // 16. Return result.
+            return Ok(String::from_str(agent, &result).into_value());
```

Comment:
**issue:** This makes an unnecessary copy, as you pass a borrow of an owned string to Nova's heap; the heap has to make a copy of the string data. You then drop that data.

You can pass ownership of an owned string to the heap using the `from_string` function instead. That way the string doesn't get dropped at all and the heap doesn't need to make that copy.